store/store.go

```diff
@@ -261,24 +264,66 @@ func (s *Store) GetModel(name string) *Model {
 
 // Reduce processes txn events into the models.
 func (s *Store) Reduce(events []core.Event) error {
-	codecActions, err := s.eventcodec.Reduce(events, s.datastore, baseKey)
+	txn, err := s.datastore.NewTransaction(false)
 	if err != nil {
 		return err
 	}
-	actions := make([]Action, len(codecActions))
-	for i, ca := range codecActions {
+	defer txn.Discard()
+
+	actions := make([]Action, len(events))
+	for i, e := range events {
+		key := baseKey.ChildString(e.Model()).ChildString(e.EntityID().String())
 		var actionType ActionType
-		switch codecActions[i].Type {
+		var ca *core.CodecResult
+		switch e.Type() {
 		case core.Create:
 			actionType = ActionCreate
+			exist, err := txn.Has(key)
+			if err != nil {
+				return err
+			}
+			if exist {
+				return errCantCreateExistingInstance
+			}
+			ca, err = s.eventcodec.Reduce(e, nil)
+			if err != nil {
+				return err
+			}
+			if err := txn.Put(key, ca.State); err != nil {
+				return fmt.Errorf("error when reducing create event: %v", err)
+			}
+			log.Debug("\tcreate operation applied")
 		case core.Save:
 			actionType = ActionSave
+			value, err := txn.Get(key)
+			if errors.Is(err, ds.ErrNotFound) {
+				return errSavingNonExistentInstance
+			}
+			if err != nil {
+				return err
+			}
+			ca, err = s.eventcodec.Reduce(e, value)
+			if err = txn.Put(key, ca.State); err != nil {
+				return err
+			}
+			log.Debug("\tsave operation applied")
 		case core.Delete:
```

Comment:
Both Create and Delete don't actually _need_ the previous state... so we _could_ probably collapse them together is this switch statement...